python/uv/private/lock.bzl

```diff
@@ -15,109 +15,242 @@
+)
+
+def lock(*, name, srcs, out, args = [], **kwargs):
     """Pin the requirements based on the src files.
 
     Differences with the current {obj}`compile_pip_requirements` rule:
-    - This is implemented in shell and uv.
+    - This is implemented as a rule that performs locking in a build action.
+    - Additionally one can use the runnable target.
+    - Uses `uv`.
     - This does not error out if the output file does not exist yet.
     - Supports transitions out of the box.
 
     Args:
         name: The name of the target to run for updating the requirements.
         srcs: The srcs to use as inputs.
         out: The output file.
-        upgrade: Tell `uv` to always upgrade the dependencies instead of
-            keeping them as they are.
-        universal: Tell `uv` to generate a universal lock file.
         args: Extra args to pass to `uv`.
         **kwargs: Extra kwargs passed to the {obj}`py_binary` rule.
     """
     pkg = native.package_name()
-    update_target = name + ".update"
+    update_target = "{}.update".format(name)
+    locker_target = "{}.run".format(name)
 
-    _args = [
-        "--custom-compile-command='bazel run //{}:{}'".format(pkg, update_target),
+    user_args = args
+    args = [
         "--generate-hashes",
         "--emit-index-url",
         "--no-strip-extras",
-        "--python=$(PYTHON3)",
-    ] + args + [
-        "$(location {})".format(src)
-        for src in srcs
     ]
-    if upgrade:
-        _args.append("--upgrade")
-    if universal:
-        _args.append("--universal")
-    _args.append("--output-file=$@")
-    cmd = "$(UV_BIN) pip compile " + " ".join(_args)
-
-    # Make a copy to ensure that we are not modifying the initial list
-    srcs = list(srcs)
-
-    # Check if the output file already exists, if yes, first copy it to the
-    # output file location in order to make `uv` not change the requirements if
-    # we are just running the command.
-    if native.glob([out]):
-        cmd = "cp -v $(location {}) $@; {}".format(out, cmd)
-        srcs.append(out)
-
-    native.genrule(
+    args += user_args
+    maybe_out = _maybe_path(out)
+
+    _lock(
         name = name,
         srcs = srcs,
-        outs = [out + ".new"],
-        cmd_bash = cmd,
+        # Check if the output file already exists, if yes, first copy it to the
+        # output file location in order to make `uv` not change the requirements if
+        # we are just running the command.
+        maybe_out = maybe_out,
+        out = out + ".new",
         tags = [
             "local",
             "manual",
             "no-cache",
         ],
+        args = args,
         target_compatible_with = _REQUIREMENTS_TARGET_COMPATIBLE_WITH,
-        toolchains = [
-            Label("//python/uv:current_toolchain"),
-            Label("//python:current_py_toolchain"),
-        ],
     )
 
+    _run_lock(
+        name = locker_target,
+        lock = name,
+    )
+
+    if maybe_out:
+        diff_test(
```

Comment:
The diff test is only added if the output file already exists in the source tree.

Maybe we should have a parameter called `no_test = False` to the macro?